tools/subtype-gen/rules.yaml

```diff
@@ -0,0 +1,433 @@
+            - subtype:
+                super: super.Type
+                sub: sub.Type
+        - subtype:
+            super: super.Type
+            sub: sub
+
+  - super: DictionaryType
+    predicate:
+      and:
+        - mustType:
+            source: sub
+            type: DictionaryType
+        - subtype:
+            super: super.ValueType
+            sub: sub.ValueType
+        - subtype:
+            super: super.KeyType
+            sub: sub.KeyType
+
+  - super: VariableSizedType
+    predicate:
+      and:
+        - mustType:
+            source: sub
+            type: VariableSizedType
+        - subtype:
+            super: super.ElementType
+            sub: sub.ElementType
+
+  - super: ConstantSizedType
+    predicate:
+      and:
+        - mustType:
+            source: sub
+            type: ConstantSizedType
+        - equals:
+            source: super.Size
+            target: sub.Size
+        - subtype:
+            super: super.ElementType
+            sub: sub.ElementType
+
+  - super: ReferenceType
+    predicate:
+      and:
+        - mustType:
+            source: sub
+            type: ReferenceType
+        - permits:
+            super: super.Authorization
+            sub: sub.Authorization
+        - subtype:
+            super: super.ReferencedType
+            sub: sub.ReferencedType
+
+  - super: CompositeType
+    predicate:
+      or:
+        - and:
+          - mustType:
+              source: sub
+              type: IntersectionType
+          - not:
+              equals:
+                source: sub.LegacyType
+                target:
+                  oneOf: [nil, AnyResourceType, AnyStructType, AnyType]
+          - mustType:
+              source: sub.LegacyType
+              type: CompositeType
+          - equals:
+              source: sub.LegacyType
+              target: super
+        - and:
+            - mustType:
+                source: sub
+                type: CompositeType
+            - never
+        - isParameterizedSubtype:
+            sub: sub
+            super: super
+
+  - super: InterfaceType
+    predicate:
+      or:
+        - and:
+          - mustType:
+              source: sub
+              type: CompositeType
+          - equals:
+              source: sub.Kind
+              target: super.CompositeKind
+          - setContains:
+              source: sub.EffectiveInterfaceConformanceSet
+              target: super
+        - and:
+          - mustType:
+              source: sub
+              type: IntersectionType
+          - setContains:
+              source: sub.EffectiveIntersectionSet
+              target: super
+        - and:
+          - mustType:
+              source: sub
+              type: InterfaceType
+          - setContains:
+              source: sub.EffectiveInterfaceConformanceSet
+              target: super
+        - isParameterizedSubtype:
+            sub: sub
+            super: super
+
+  - super: IntersectionType
```

Comment:
Maybe as a TODO before merging the feature branch: Review and verify these and the other rules, as well as the code that is generated for them, and compare to the original Go code.

Reply:
Added an item here: https://github.com/onflow/cadence/issues/3691#issuecomment-3439450533